src/mesh.cpp

```diff
@@ -220,20 +220,55 @@ void UnstructuredMesh::to_hdf5(hid_t group) const
   write_dataset(mesh_group, "type", mesh_type);
   write_dataset(mesh_group, "filename", filename_);
   write_dataset(mesh_group, "library", this->library());
-  // write volume of each element
-  vector<double> tet_vols;
-  xt::xtensor<double, 2> centroids({static_cast<size_t>(this->n_bins()), 3});
+
+  if (specified_length_multiplier_)
+    write_dataset(mesh_group, "length_multiplier", length_multiplier_);
+
+  // write vertex coordinates
+  xt::xtensor<double, 2> vertices({static_cast<size_t>(this->n_vertices()), 3});
+  for (int i = 0; i < this->n_vertices(); i++) {
+    auto v = this->vertex(i);
+    xt::view(vertices, i, xt::all()) = xt::xarray<double>({v.x, v.y, v.z});
+  }
+  write_dataset(mesh_group, "vertices", vertices);
+
+  int num_elem_skipped = 0;
+
+  // write element types and connectivity
+  vector<double> volumes;
+  xt::xtensor<int, 2> connectivity ({static_cast<size_t>(this->n_bins()), 8});
+  xt::xtensor<int, 2> elem_types ({static_cast<size_t>(this->n_bins()), 1});
   for (int i = 0; i < this->n_bins(); i++) {
-    tet_vols.emplace_back(this->volume(i));
-    auto c = this->centroid(i);
-    xt::view(centroids, i, xt::all()) = xt::xarray<double>({c.x, c.y, c.z});
+    auto conn = this->connectivity(i);
+
+    volumes.emplace_back(this->volume(i));
+
+    // write linear tet element
+    if (conn.size() == 4) {
+      xt::view(elem_types, i, xt::all()) = static_cast<int>(ElementType::LINEAR_TET);
+      xt::view(connectivity, i, xt::all()) = xt::xarray<int>({conn[0], conn[1], conn[2], conn[3],
+                                                              -1, -1, -1, -1});
+    // write linear hex element
+    } else if (conn.size() == 8) {
+      xt::view(elem_types, i, xt::all()) = static_cast<int>(ElementType::LINEAR_HEX);
+      xt::view(connectivity, i, xt::all()) = xt::xarray<int>({conn[0], conn[1], conn[2], conn[3],
+                                                              conn[4], conn[5], conn[6], conn[7]});
+    } else {
+      num_elem_skipped++;
+      xt::view(elem_types, i, xt::all()) = static_cast<int>(ElementType::UNSUPPORTED);
+      xt::view(connectivity, i, xt::all()) = xt::xarray<int>({-1, -1, -1, -1, -1, -1, -1, -1});
```

Comment:
If you assign a scalar, will it broadcast to all? i.e. `= -1`

Reply:
Looks like it does! Much cleaner. Thank you.